lib/puppetserver/ca/action/generate_csr.rb

```diff
@@ -0,0 +1,123 @@
+require 'optparse'
+
+require 'puppetserver/ca/config/puppet'
+require 'puppetserver/ca/errors'
+require 'puppetserver/ca/local_certificate_authority'
+require 'puppetserver/ca/utils/cli_parsing'
+require 'puppetserver/ca/utils/file_system'
+require 'puppetserver/ca/utils/signing_digest'
+
+module Puppetserver
+  module Ca
+    module Action
+      class GenerateCsr
+        include Puppetserver::Ca::Utils
+
+        SUMMARY = 'Generate a CSR for an intermediate signing CA cert'
+        BANNER = <<-BANNER
+Usage:
+  puppetserver ca generate-csr [--help]
+  puppetserver ca generate-csr --output-dir PATH [--config PATH] [--ca-name NAME]
+
+Description:
+  Setup a root and intermediate signing CA for Puppet Server
+  and store generated CA keys, certs, crls, and associated
+  master related files on disk.
+
+Options:
+        BANNER
+
+        def initialize(logger)
+          @logger = logger
+        end
+
+        def run(input)
+          # Validate config_path provided
+          config_path = input['config']
+          if config_path
+            errors = FileSystem.validate_file_paths(config_path)
+            return 1 if Errors.handle_with_usage(@logger, errors)
+          end
+
+          # Load, resolve, and validate puppet config settings
+          settings_overrides = {}
+          settings_overrides[:ca_name] = input['ca-name'] unless input['ca-name'].empty?
+
+          puppet = Config::Puppet.new(config_path)
+          puppet.load(settings_overrides)
+          return 1 if Errors.handle_with_usage(@logger, puppet.errors)
+
+          # Load most secure signing digest we can for cers/crl/csr signing.
+          signer = SigningDigest.new
+          return 1 if Errors.handle_with_usage(@logger, signer.errors)
+
+          # Generate intermediate ca and put the results in the specified output directory
+          generate_csr(puppet.settings, signer.digest, input['output-dir'])
+          return 0
+        end
+
+        def generate_csr(settings, signing_digest, dir)
+          ca = Puppetserver::Ca::LocalCertificateAuthority.new(signing_digest, settings)
+
+          key, csr = ca.create_intermediate_csr
+
+          write_generated_files(key, csr, dir)
+        end
+
+        def write_generated_files(key, csr, dir)
+          FileSystem.write_file(File.join(dir, 'ca.key'), key.to_s, 0640)
+          FileSystem.write_file(File.join(dir, 'ca.csr'), csr.to_s, 0640)
+        end
+
+        def check_for_existing_ssl_files(dir)
+          files = [ File.join(dir, 'ca.key'), File.join(dir, 'ca.csr')]
+          errors = Puppetserver::Ca::Utils::FileSystem.check_for_existing_files(files)
+          if !errors.empty?
+            errors << 'Please delete these files if you want to generate a CSR'
```

Comment:
what files?

Reply:
I based that on `action/generate.rb` which does the same thing. 

`Puppetserver::Ca::Utils::FileSystem.check_for_existing_files` adds `"Existing file at '#{path}'"` to `errors` for each file that already exists from the list, so those messages will also present, which is why this error message doesn't specify them. There's also a test that checks that stderr contains `ca.csr` and `ca.key` in the case of existing files.